pkg/ebpf/tcp_queue_length.go

```diff
@@ -0,0 +1,152 @@
+	// this would also require to properly handle inclusion guards.
+	includeRegexp := regexp.MustCompile(`^\s*#\s*include\s+"(.*)"$`)
+	var source bytes.Buffer
+	scanner := bufio.NewScanner(bytes.NewBuffer(source_raw))
+	for scanner.Scan() {
+		match := includeRegexp.FindSubmatch(scanner.Bytes())
+		if len(match) == 2 {
+			header, err := Asset(string(match[1]))
+			if err == nil {
+				source.Write(header)
+				continue
+			}
+		}
+		source.Write(scanner.Bytes())
+		source.WriteByte('\n')
+	}
+
+	m := bpflib.NewModule(source.String(), []string{})
+	if m == nil {
+		return nil, fmt.Errorf("Failed to compile “tcp-queue-length-kern.c”")
+	}
+
+	kprobe_recvmsg, err := m.LoadKprobe("kprobe__tcp_recvmsg")
+	if err != nil {
+		return nil, fmt.Errorf("Failed to load kprobe__tcp_recvmsg: %s\n", err)
+	}
+
+	if err := m.AttachKprobe("tcp_recvmsg", kprobe_recvmsg, -1); err != nil {
+		return nil, fmt.Errorf("Failed to attach tcp_recvmsg: %s\n", err)
+	}
+
+	kretprobe_recvmsg, err := m.LoadKprobe("kretprobe__tcp_recvmsg")
+	if err != nil {
+		return nil, fmt.Errorf("Failed to load kretprobe__tcp_recvmsg: %s\n", err)
+	}
+
+	if err := m.AttachKretprobe("tcp_recvmsg", kretprobe_recvmsg, -1); err != nil {
+		return nil, fmt.Errorf("Failed to attach tcp_recvmsg: %s\n", err)
+	}
+
+	kprobe_sendmsg, err := m.LoadKprobe("kprobe__tcp_sendmsg")
+	if err != nil {
+		return nil, fmt.Errorf("Failed to load kprobe__tcp_recvmsg: %s\n", err)
+	}
+
+	if err := m.AttachKprobe("tcp_sendmsg", kprobe_sendmsg, -1); err != nil {
+		return nil, fmt.Errorf("Failed to attach tcp_sendmsg: %s\n", err)
+	}
+
+	kretprobe_sendmsg, err := m.LoadKprobe("kretprobe__tcp_sendmsg")
+	if err != nil {
+		return nil, fmt.Errorf("Failed to load kretprobe__tcp_recvmsg: %s\n", err)
+	}
+
+	if err := m.AttachKretprobe("tcp_sendmsg", kretprobe_sendmsg, -1); err != nil {
+		return nil, fmt.Errorf("Failed to attach tcp_sendmsg: %s\n", err)
+	}
+
+	table := bpflib.NewTable(m.TableId("queue"), m)
+
+	return &TCPQueueLengthTracer{
+		m:        m,
+		queueMap: table,
+	}, nil
+}
+
+func (t *TCPQueueLengthTracer) Close() {
+	t.m.Close()
+}
+
+func (t *TCPQueueLengthTracer) Get() []Stats {
+	if t == nil {
+		return nil
+	}
+
+	var result []Stats
+
+	for it := t.queueMap.Iter(); it.Next(); {
+		var in C.struct_stats // kernel       <-> system-probe
+		var out Stats         // system-probe <-> agent
+
+		// `binary.Read(…)` doesn’t work because reflection doesn’t work with C types.
+		// binary.Read(bytes.NewBuffer(it.Leaf()), bpflib.GetHostByteOrder(), &in)
+
+		data := it.Leaf()
+		C.memcpy(unsafe.Pointer(&in), unsafe.Pointer(&data[0]), C.sizeof_struct_stats)
+
+		// TODO: Can this code be handled by using reflection? Would it be clearer?
+		out.Pid = uint32(in.pid)
+		out.ContainerID = C.GoString(&in.cgroup_name[0])
+		out.Conn.Saddr = make(net.IP, 4)
+		bpflib.GetHostByteOrder().PutUint32(out.Conn.Saddr, uint32(in.conn.saddr))
```

Comment:
Interesting, I think we can get rid of some code in `ebpf/common.go` using this (which detects the endianness of the host).
Also, we could have a global variable `ebpf.NativeEndianess`, but I can do that in my refactor PR.